File: pipecut/planner.py
"""Ước lượng thời gian chạy **đúng như FluidNC lập kế hoạch chuyển động**.

Mô hình trong :mod:`pipecut.gsim` coi máy chạy đều theo F, không tăng giảm tốc.
Đủ để vẽ mô phỏng, nhưng nó *không thấy* đúng cái làm máy thật chậm: mỗi lần
dừng hẳn rồi khởi động lại.  Nhấc Z, chạy ngang, hạ Z thành ba lệnh riêng thì
máy phải dừng hẳn ở hai chỗ; mỗi góc gấp trên đường cắt cũng phải hãm lại.
Tối ưu chuyển động mà không đo được mấy cái đó thì chỉ là đoán.

Module này chạy lại đúng thuật toán lập kế hoạch của Grbl/FluidNC:

* **Tốc độ danh định** của mỗi khối: F (hoặc tốc độ chạy nhanh với G0), nhưng
  không trục nào được vượt ``max_rate`` của nó.
* **Gia tốc** của khối: trục nào yếu nhất theo đúng hướng đi quyết định.
* **Tốc độ qua điểm nối** giữa hai khối theo công thức *junction deviation*:
  góc càng gấp thì càng phải chậm, quay ngược đầu thì phải dừng hẳn.
* **Nhìn trước có hạn**: bộ đệm chỉ chứa ``planner_blocks`` khối, khối cuối
  trong bộ đệm luôn phải dừng được - nên chuỗi đoạn quá ngắn tự làm máy chậm.
* **Lệnh đồng bộ** (M3/M5, G4, G10...) bắt bộ đệm chạy cạn: dừng hẳn.

Mỗi khối được xếp vào một loại chuyển động (cắt, chạy không, nhấc/hạ, ...)
để biết thời gian đang tiêu vào đâu.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Sequence, Tuple

from .config import MachineProfile, ROLE_RADIAL
from .gcode import strip_gcode_comment

CATEGORIES: Tuple[str, ...] = ("cut", "travel", "lift", "plunge", "index", "dwell")
CATEGORY_LABELS: Dict[str, str] = {
    "cut": "cắt",
    "travel": "chạy không",
    "lift": "nhấc/hạ mỏ",
    "plunge": "hạ mỏ chậm",
    "index": "xoay/đi chậm không cắt",
    "dwell": "chờ (mồi, tắt)",
}

# Lệnh làm bộ đệm chạy cạn rồi mới làm tiếp (Grbl: protocol_buffer_synchronize)
_SYNC_M = {3, 4, 5, 7, 8, 9, 62, 63, 64, 65}
_SYNC_G = {4, 10, 28, 30, 92, 53}


@dataclass
class Block:
    """Một đoạn chuyển động thẳng trong không gian trục (mm và độ như nhau)."""

    line: int
    length: float
    unit: Dict[str, float]       # véc-tơ đơn vị hướng đi
    v_nom: float                 # mm/s
    accel: float                 # mm/s^2
    category: str
    sync_before: bool = False    # có lệnh đồng bộ ngay trước khối này
    v_junction: float = 0.0      # tốc độ tối đa được phép ở điểm vào (mm/s)
    v_entry: float = 0.0
    v_exit: float = 0.0
    time: float = 0.0


@dataclass
class PlanResult:
    """Kết quả lập kế hoạch cho cả chương trình."""

    total: float = 0.0
    by_category: Dict[str, float] = field(default_factory=lambda: {c: 0.0 for c in CATEGORIES})
    distance: Dict[str, float] = field(default_factory=lambda: {c: 0.0 for c in CATEGORIES})
    blocks: int = 0
    syncs: int = 0               # số lần bộ đệm bị bắt chạy cạn
    full_stops: int = 0          # số điểm nối phải dừng hẳn (không tính đồng bộ)
    nominal: float = 0.0         # thời gian nếu chạy đều theo F, không tăng giảm tốc
    line_times: Dict[int, float] = field(default_factory=dict)   # số dòng -> giây

    @property
    def accel_overhead(self) -> float:
        """Thời gian mất thêm vì phải tăng/giảm tốc so với chạy đều."""
        return max(0.0, self.total - self.nominal - self.by_category.get("dwell", 0.0))

    def share(self, category: str) -> float:
        return self.by_category.get(category, 0.0) / self.total if self.total > 0 else 0.0

    def summary(self) -> str:
        parts = [f"{CATEGORY_LABELS[c]} {self.by_category[c]:.1f}s"
                 for c in CATEGORIES if self.by_category[c] > 0.05]
        return (f"tổng {self.total:.1f}s ({', '.join(parts)}); "
                f"{self.full_stops} lần dừng hẳn giữa chừng, {self.syncs} lần đồng bộ")


# ----------------------------------------------------------------------
# Đọc chương trình thành các khối
# ----------------------------------------------------------------------
def _words(text: str) -> List[Tuple[str, float]]:
    out: List[Tuple[str, float]] = []
    s = text.upper()
    i, n = 0, len(s)
    while i < n:
        ch = s[i]
        if "A" <= ch <= "Z":
            j = i + 1
            while j < n and (s[j].isdigit() or s[j] in "+-."):
                j += 1
            try:
                out.append((ch, float(s[i + 1:j])))
            except ValueError:
                pass
            i = j
        else:
            i += 1
    return out


def _limit_by_axes(limits: Dict[str, float], unit: Dict[str, float]) -> float:
    """Giống ``limit_value_by_axis_maximum`` của Grbl."""
    best = float("inf")
    for letter, u in unit.items():
        if abs(u) > 1e-12:
            lim = limits.get(letter, 0.0)
            if lim > 0:
                best = min(best, lim / abs(u))
    return best if best < float("inf") else 0.0


def read_blocks(profile: MachineProfile, lines: Sequence[str]) -> List[Block]:
    letters = [c for c in profile.letters]
    rates = {}      # mm/s
    accels = {}     # mm/s^2
    for c in letters:
        ax = profile.axis_by_letter(c)
        rates[c] = (ax.max_rate if ax and ax.max_rate > 0 else 3000.0) / 60.0
        accels[c] = ax.accel if ax and ax.accel > 0 else 200.0
    radial = (profile.letter(ROLE_RADIAL) or "Z").upper()

    # Lúc bắt đầu, mỏ đứng ở gốc chi tiết - đúng chỗ vừa "đặt gốc tại đây".
    # Cùng giả định với mô phỏng (gsim) và bộ sắp thứ tự cắt (jobs).
    pos = {c: 0.0 for c in letters}
    rapid, absolute, torch = True, True, False
    feed = 0.0
    sync_pending = True    # đầu chương trình: máy đứng yên
    blocks: List[Block] = []

    for index, raw in enumerate(lines, start=1):
        text = strip_gcode_comment(raw)
        if not text:
            continue
        words = _words(text)
        target = dict(pos)
        moved = False
        dwell = None
        is_dwell = False
        for letter, value in words:
            if letter == "G":
                code = int(round(value * 10))
                if code == 0:
                    rapid = True
                elif code == 10:
                    rapid = False
                elif code == 900:
                    absolute = True
                elif code == 910:
                    absolute = False
                elif code == 40:
                    is_dwell = True
                if code % 10 == 0 and code // 10 in _SYNC_G:
                    sync_pending = True
            elif letter == "M":
                code = int(round(value))
                if code in (3, 4):
                    torch = True
                elif code in (5, 2, 30):
                    torch = False
                if code in _SYNC_M:
                    sync_pending = True
            elif letter == "F":
                feed = max(1.0, value)
            elif letter == "P" and is_dwell:
                dwell = max(0.0, value)
            elif letter in pos:
                target[letter] = value if absolute else pos[letter] + value
                moved = True

        if is_dwell:
            if dwell:
                blocks.append(Block(index, 0.0, {}, 0.0, 1.0, "dwell",
                                    sync_before=True, time=dwell))
            sync_pending = True
            continue
        if not moved:
            continue
        delta = {c: target[c] - pos[c] for c in letters if abs(target[c] - pos[c]) > 1e-9}
        length = math.sqrt(sum(d * d for d in delta.values()))
        if length < 1e-9:
            pos = target
            continue
        unit = {c: d / length for c, d in delta.items()}
        rapid_rate = _limit_by_axes(rates, unit)
        v_nom = rapid_rate if rapid else min(feed / 60.0, rapid_rate)
        only_z = set(delta) == {radial}
        if rapid:
            category = "lift" if only_z else "travel"
        elif only_z:
            category = "plunge"      # hạ từ độ cao mồi xuống độ cao cắt
        elif torch:
            category = "cut"
        else:
            category = "index"
        blocks.append(Block(index, length, unit, max(v_nom, 1e-6),
                            max(_limit_by_axes(accels, unit), 1e-6), category,
                            sync_before=sync_pending))
        sync_pending = False
        pos = target
    _junctions(blocks, profile, accels)
    return blocks


def _junctions(blocks: List[Block], profile: MachineProfile, accels: Dict[str, float]) -> None:
    """Tốc độ tối đa tại điểm vào mỗi khối - công thức junction deviation."""
    jd = max(1e-6, float(getattr(profile.motion, "junction_deviation", 0.01)))
    prev: Optional[Block] = None
    for b in blocks:
        if b.category == "dwell":
            prev = None
            continue
        if prev is None or b.sync_before:
            b.v_junction = 0.0
        else:
            letters = set(b.unit) | set(prev.unit)
            cos_t = -sum(prev.unit.get(c, 0.0) * b.unit.get(c, 0.0) for c in letters)
            if cos_t > 0.999999:            # quay ngược đầu: phải dừng hẳn
                v2 = 0.0
            elif cos_t < -0.999999:         # thẳng hàng: không giới hạn
                v2 = float("inf")
            else:
                ju = {c: b.unit.get(c, 0.0) - prev.unit.get(c, 0.0) for c in letters}
                n = math.sqrt(sum(v * v for v in ju.values()))
                ju = {c: v / n for c, v in ju.items()} if n > 1e-12 else ju
                a_j = _limit_by_axes(accels, ju) or b.accel
                s = math.sqrt(0.5 * (1.0 - cos_t))
                v2 = a_j * jd * s / (1.0 - s)
            b.v_junction = math.sqrt(min(v2, b.v_nom ** 2, prev.v_nom ** 2))
        prev = b


def _trapezoid(length: float, v0: float, v1: float, vc: float, a: float) -> float:
    if length <= 0:
        return 0.0
    d_acc = max(0.0, (vc * vc - v0 * v0) / (2 * a))
    d_dec = max(0.0, (vc * vc - v1 * v1) / (2 * a))
    if d_acc + d_dec <= length:
        return (vc - v0) / a + (vc - v1) / a + (length - d_acc - d_dec) / vc
    vp = math.sqrt(max(0.0, (2 * a * length + v0 * v0 + v1 * v1) / 2.0))
    vp = max(vp, v0, v1)
    return max(0.0, (vp - v0) / a + (vp - v1) / a)


def plan(profile: MachineProfile, lines: Sequence[str]) -> PlanResult:
    """Lập kế hoạch cả chương trình, trả về thời gian và phân loại."""
    blocks = read_blocks(profile, lines)
    res = PlanResult()
    depth = max(2, int(getattr(profile.motion, "planner_blocks", 32)) - 1)

    # Chia thành các đoạn liền mạch giữa hai lần đồng bộ.
    runs: List[List[Block]] = []
    cur: List[Block] = []
    for b in blocks:
        if b.category == "dwell":
            if cur:
                runs.append(cur)
                cur = []
            res.by_category["dwell"] += b.time
            res.total += b.time
            res.line_times[b.line] = res.line_times.get(b.line, 0.0) + b.time
            continue
        if b.sync_before and cur:
            runs.append(cur)
            cur = []
        cur.append(b)
    if cur:
        runs.append(cur)
    res.syncs = len(runs)

    for run in runs:
        n = len(run)
        # quãng hãm có được trong phần còn lại của bộ đệm sau mỗi khối
        room = [0.0] * n
        pref = [0.0] * (n + 1)
        for i, b in enumerate(run):
            pref[i + 1] = pref[i] + 2.0 * b.accel * b.length
        for i in range(n):
            j = min(n, i + depth)
            room[i] = pref[j] - pref[i + 1]
        # lượt ngược: phải hãm kịp trước mọi giới hạn phía trước
        v_next = 0.0
        exits = [0.0] * n
        for i in range(n - 1, -1, -1):
            b = run[i]
            exits[i] = min(v_next, math.sqrt(max(0.0, room[i])))
            v_next = min(b.v_junction,
                         math.sqrt(exits[i] ** 2 + 2.0 * b.accel * b.length))
        # lượt xuôi: không tăng tốc nhanh hơn gia tốc cho phép
        v_in = 0.0
        for i, b in enumerate(run):
            b.v_entry = min(v_in, b.v_junction) if i else 0.0
            b.v_exit = min(exits[i], math.sqrt(b.v_entry ** 2 + 2.0 * b.accel * b.length), b.v_nom)
            b.time = _trapezoid(b.length, b.v_entry, b.v_exit, b.v_nom, b.accel)
            v_in = b.v_exit
            res.by_category[b.category] += b.time
            res.line_times[b.line] = res.line_times.get(b.line, 0.0) + b.time
            res.distance[b.category] += b.length
            res.total += b.time
            res.nominal += b.length / b.v_nom
            res.blocks += 1
            if i and b.v_entry < 0.05 * min(b.v_nom, run[i - 1].v_nom):
                res.full_stops += 1
    return res

File: c/src/planner.c
/* Lập kế hoạch chuyển động đúng như Grbl/FluidNC - chuyển từ pipecut/planner.py.
 *   - tốc độ danh định: F (hoặc tốc độ chạy nhanh với G0), không trục nào vượt max_rate
 *   - gia tốc khối: trục yếu nhất theo đúng hướng đi
 *   - tốc độ qua điểm nối: công thức junction deviation
 *   - bộ đệm nhìn trước có hạn (planner_blocks), lệnh M3/M5/G4/G10 bắt dừng hẳn */
#include <ctype.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include "core.h"

const char *CAT_NAMES_VI[NCAT] = {"cắt", "chạy không", "nhấc/hạ mỏ", "hạ mỏ chậm",
                                  "xoay/đi chậm", "chờ (mồi, tắt)"};

typedef struct {
    int line;
    double len, unit[NAX], v_nom, accel, v_junc;
    int cat, sync, dwell;
    double dwell_t;
    double a[NAX], b[NAX];
    int torch, rapid;
} Blk;

void plan_free(Plan *p) { free(p->mv); memset(p, 0, sizeof *p); }

static double limit_by_axes(const double lim[NAX], const double u[NAX])
{
    double best = 1e300;
    for (int i = 0; i < NAX; i++)
        if (fabs(u[i]) > 1e-12 && lim[i] > 0 && lim[i] / fabs(u[i]) < best) best = lim[i] / fabs(u[i]);
    return best < 1e299 ? best : 0.0;
}

static double trap_time(double L, double v0, double v1, double vc, double a)
{
    if (L <= 0) return 0;
    double da = (vc * vc - v0 * v0) / (2 * a), dd = (vc * vc - v1 * v1) / (2 * a);
    if (da < 0) da = 0;
    if (dd < 0) dd = 0;
    if (da + dd <= L) return (vc - v0) / a + (vc - v1) / a + (L - da - dd) / vc;
    double vp = sqrt((2 * a * L + v0 * v0 + v1 * v1) / 2.0);
    if (vp < v0) vp = v0;
    if (vp < v1) vp = v1;
    double t = (vp - v0) / a + (vp - v1) / a;
    return t > 0 ? t : 0;
}

/* quãng đường đã đi sau thời gian t trong khối hình thang */
static double trap_dist(double t, double L, double v0, double v1, double vc, double a)
{
    double da = (vc * vc - v0 * v0) / (2 * a), dd = (vc * vc - v1 * v1) / (2 * a);
    if (da < 0) da = 0;
    if (dd < 0) dd = 0;
    double vp = vc;
    if (da + dd > L) {
        vp = sqrt((2 * a * L + v0 * v0 + v1 * v1) / 2.0);
        da = (vp * vp - v0 * v0) / (2 * a);
        dd = (vp * vp - v1 * v1) / (2 * a);
    }
    double ta = (vp - v0) / a, tc = (L - da - dd) / (vp > 0 ? vp : 1);
    if (tc < 0) tc = 0;
    if (t <= ta) return v0 * t + 0.5 * a * t * t;
    if (t <= ta + tc) return da + vp * (t - ta);
    double td = t - ta - tc;
    double s = da + vp * tc + vp * td - 0.5 * a * td * td;
    return s > L ? L : s;
}

int plan_program(const char *gcode, const Machine *m, Plan *out)
{
    memset(out, 0, sizeof *out);
    double rates[NAX], acc[NAX];
    for (int i = 0; i < NAX; i++) { rates[i] = m->max_rate[i] / 60.0; acc[i] = m->accel[i]; }

    int cap = 1024, nb = 0;
    Blk *bl = (Blk *)malloc((size_t)cap * sizeof *bl);
    double pos[NAX] = {0, 0, 0, 0};
    int rapid = 1, absolute = 1, torch = 0, sync_pending = 1;
    double feed = 0;
    int line_no = 0;
    const char *p = gcode;
    char buf[512];
    while (*p) {
        size_t k = 0;
        while (*p && *p != '\n' && k < sizeof buf - 1) buf[k++] = *p++;
        buf[k] = 0;
        while (*p && *p != '\n') p++;
        if (*p == '\n') p++;
        line_no++;
        /* bỏ chú thích (...) và ;... */
        char txt[512];
        size_t t = 0;
        int paren = 0;
        for (size_t i = 0; buf[i]; i++) {
            char ch = buf[i];
            if (ch == '(') { paren = 1; continue; }
            if (ch == ')') { paren = 0; continue; }
            if (ch == ';') break;
            if (!paren) txt[t++] = (char)toupper((unsigned char)ch);
        }
        txt[t] = 0;
        double tgt[NAX];
        memcpy(tgt, pos, sizeof tgt);
        int moved = 0, is_dwell = 0;
        double dwell = -1;
        for (size_t i = 0; i < t;) {
            char L = txt[i];
            if (L < 'A' || L > 'Z') { i++; continue; }
            char *end;
            double val = strtod(txt + i + 1, &end);
            if (end == txt + i + 1) { i++; continue; }
            i = (size_t)(end - txt);
            if (L == 'G') {
                int code = (int)floor(val * 10 + 0.5);
                if (code == 0) rapid = 1;
                else if (code == 10) rapid = 0;
                else if (code == 900) absolute = 1;
                else if (code == 910) absolute = 0;
                else if (code == 40) is_dwell = 1;
                if (code % 10 == 0) {
                    int g = code / 10;
                    if (g == 4 || g == 10 || g == 28 || g == 30 || g == 92 || g == 53) sync_pending = 1;
                }
            } else if (L == 'M') {
                int code = (int)floor(val + 0.5);
                if (code == 3 || code == 4) torch = 1;
                else if (code == 5 || code == 2 || code == 30) torch = 0;
                if (code == 3 || code == 4 || code == 5 || code == 7 || code == 8 || code == 9 ||
                    (code >= 62 && code <= 65)) sync_pending = 1;
            } else if (L == 'F') {
                feed = val > 1 ? val : 1;
            } else if (L == 'P' && is_dwell) {
                dwell = val > 0 ? val : 0;
            } else {
                int ax = L == 'X' ? AX_X : L == 'Y' ? AX_Y : L == 'Z' ? AX_Z : L == 'A' ? AX_A : -1;
                if (ax >= 0) { tgt[ax] = absolute ? val : pos[ax] + val; moved = 1; }
            }
        }
        if (nb + 1 >= cap) { cap *= 2; bl = (Blk *)realloc(bl, (size_t)cap * sizeof *bl); }
        if (is_dwell) {
            if (dwell > 0) {
                Blk *b = &bl[nb++];
                memset(b, 0, sizeof *b);
                b->line = line_no; b->dwell = 1; b->dwell_t = dwell; b->sync = 1; b->cat = CAT_DWELL;
                b->torch = torch;
                memcpy(b->a, pos, sizeof pos); memcpy(b->b, pos, sizeof pos);
            }
            sync_pending = 1;
            continue;
        }
        if (!moved) continue;
        double d[NAX], len = 0;
        for (int i = 0; i < NAX; i++) { d[i] = tgt[i] - pos[i]; len += d[i] * d[i]; }
        len = sqrt(len);
        if (len < 1e-9) { memcpy(pos, tgt, sizeof pos); continue; }
        Blk *b = &bl[nb++];
        memset(b, 0, sizeof *b);
        b->line = line_no; b->len = len;
        for (int i = 0; i < NAX; i++) b->unit[i] = d[i] / len;
        double rr = limit_by_axes(rates, b->unit);
        b->v_nom = rapid ? rr : (feed / 60.0 < rr ? feed / 60.0 : rr);
        if (b->v_nom < 1e-6) b->v_nom = 1e-6;
        b->accel = limit_by_axes(acc, b->unit);
        if (b->accel < 1e-6) b->accel = 1e-6;
        int only_z = fabs(d[AX_X]) < 1e-9 && fabs(d[AX_Y]) < 1e-9 && fabs(d[AX_A]) < 1e-9;
        b->cat = rapid ? (only_z ? CAT_LIFT : CAT_TRAVEL)
                       : (only_z ? CAT_PLUNGE : (torch ? CAT_CUT : CAT_INDEX));
        b->sync = sync_pending;
        b->torch = torch; b->rapid = rapid;
        memcpy(b->a, pos, sizeof pos); memcpy(b->b, tgt, sizeof tgt);
        sync_pending = 0;
        memcpy(pos, tgt, sizeof pos);
    }

    /* tốc độ tối đa tại điểm vào mỗi khối - junction deviation */
    double jd = m->junction_deviation > 1e-6 ? m->junction_deviation : 1e-6;
    Blk *prev = NULL;
    for (int i = 0; i < nb; i++) {
        Blk *b = &bl[i];
        if (b->dwell) { prev = NULL; continue; }
        if (!prev || b->sync) { b->v_junc = 0; }
        else {
            double cs = 0;
            for (int k = 0; k < NAX; k++) cs -= prev->unit[k] * b->unit[k];
            double v2;
            if (cs > 0.999999) v2 = 0;
            else if (cs < -0.999999) v2 = 1e300;
            else {
                double ju[NAX], n = 0;
                for (int k = 0; k < NAX; k++) { ju[k] = b->unit[k] - prev->unit[k]; n += ju[k] * ju[k]; }
                n = sqrt(n);
                if (n > 1e-12) for (int k = 0; k < NAX; k++) ju[k] /= n;
                double aj = limit_by_axes(acc, ju);
                if (aj <= 0) aj = b->accel;
                double sh = sqrt(0.5 * (1 - cs));
                v2 = aj * jd * sh / (1 - sh);
            }
            double lim = b->v_nom * b->v_nom < prev->v_nom * prev->v_nom ? b->v_nom * b->v_nom : prev->v_nom * prev->v_nom;
            b->v_junc = sqrt(v2 < lim ? v2 : lim);
        }
        prev = b;
    }

    /* các đoạn liền mạch giữa hai lần đồng bộ */
    out->cap = nb + 1;
    out->mv = (Move *)calloc((size_t)out->cap, sizeof(Move));
    int depth = (m->planner_blocks > 2 ? m->planner_blocks : 2) - 1;
    double tnow = 0;
    int i = 0;
    while (i < nb) {
        if (bl[i].dwell) {
            Move *mv = &out->mv[out->n++];
            mv->t0 = tnow; mv->dt = bl[i].dwell_t; mv->dwell = 1; mv->line = bl[i].line;
            mv->torch = bl[i].torch;
            memcpy(mv->a, bl[i].a, sizeof mv->a); memcpy(mv->b, bl[i].b, sizeof mv->b);
            out->by_cat[CAT_DWELL] += mv->dt;
            tnow += mv->dt;
            i++;
            continue;
        }
        int j = i + 1;
        while (j < nb && !bl[j].dwell && !bl[j].sync) j++;
        int n = j - i;
        out->syncs++;
        double *pref = (double *)calloc((size_t)n + 1, sizeof(double));
        double *ex = (double *)calloc((size_t)n, sizeof(double));
        for (int k = 0; k < n; k++) pref[k + 1] = pref[k] + 2 * bl[i + k].accel * bl[i + k].len;
        double v_next = 0;
        for (int k = n - 1; k >= 0; k--) {
            int e = k + depth < n ? k + depth : n;
            double room = pref[e] - pref[k + 1];
            double r = sqrt(room > 0 ? room : 0);
            ex[k] = v_next < r ? v_next : r;
            Blk *b = &bl[i + k];
            double reach = sqrt(ex[k] * ex[k] + 2 * b->accel * b->len);
            v_next = b->v_junc < reach ? b->v_junc : reach;
        }
        double v_in = 0;
        for (int k = 0; k < n; k++) {
            Blk *b = &bl[i + k];
            double ve = k ? (v_in < b->v_junc ? v_in : b->v_junc) : 0;
            double reach = sqrt(ve * ve + 2 * b->accel * b->len);
            double vx = ex[k];
            if (reach < vx) vx = reach;
            if (b->v_nom < vx) vx = b->v_nom;
            double dt = trap_time(b->len, ve, vx, b->v_nom, b->accel);
            Move *mv = &out->mv[out->n++];
            mv->t0 = tnow; mv->dt = dt; mv->line = b->line;
            mv->rapid = b->rapid; mv->torch = b->torch;
            mv->v_entry = ve; mv->v_exit = vx; mv->v_nom = b->v_nom; mv->accel = b->accel;
            memcpy(mv->a, b->a, sizeof mv->a); memcpy(mv->b, b->b, sizeof mv->b);
            out->by_cat[b->cat] += dt;
            tnow += dt;
            if (k && ve < 0.05 * (b->v_nom < bl[i + k - 1].v_nom ? b->v_nom : bl[i + k - 1].v_nom))
                out->full_stops++;
            v_in = vx;
        }
        free(pref);
        free(ex);
        i = j;
    }
    out->total = tnow;
    free(bl);
    return 0;
}

void plan_state(const Plan *p, double t, double pos[NAX], int *torch, int *line)
{
    if (!p->n) { memset(pos, 0, sizeof(double) * NAX); *torch = 0; *line = 0; return; }
    if (t <= 0) { memcpy(pos, p->mv[0].a, sizeof(double) * NAX); *torch = 0; *line = p->mv[0].line; return; }
    /* tìm nhị phân khối đang chạy */
    int lo = 0, hi = p->n - 1;
    while (lo < hi) {
        int mid = (lo + hi + 1) / 2;
        if (p->mv[mid].t0 <= t) lo = mid; else hi = mid - 1;
    }
    const Move *mv = &p->mv[lo];
    double f = 1.0;
    if (mv->dwell) f = 1.0;
    else if (mv->dt > 1e-12 && t < mv->t0 + mv->dt) {
        double L = 0;
        for (int i = 0; i < NAX; i++) L += (mv->b[i] - mv->a[i]) * (mv->b[i] - mv->a[i]);
        L = sqrt(L);
        f = L > 0 ? trap_dist(t - mv->t0, L, mv->v_entry, mv->v_exit, mv->v_nom, mv->accel) / L : 1.0;
        if (f > 1) f = 1;
        if (f < 0) f = 0;
    }
    for (int i = 0; i < NAX; i++) pos[i] = mv->a[i] + (mv->b[i] - mv->a[i]) * f;
    *torch = mv->torch;
    *line = mv->line;
}
